experimental/sql/src/alg_to_ssa.py

```diff
@@ -0,0 +1,151 @@
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+from dataclasses import dataclass
+from xdsl.ir import Operation, MLContext, Region, Block
+from typing import List, Type, Optional
+from xdsl.dialects.builtin import ArrayAttr, StringAttr, ModuleOp, IntegerAttr
+
+from xdsl.pattern_rewriter import RewritePattern, GreedyRewritePatternApplier, PatternRewriteWalker, PatternRewriter, op_type_rewrite_pattern
+
+import dialects.rel_alg as RelAlg
+import dialects.rel_ssa as RelSSA
+
+# This file contains the rewrite infrastructure to translate the relational
+# algebra dialect to the relational SSA dialect. The current design has a parent
+# class `RelAlgRewriter` that contains functions used for several `Rewriter`s.
+# All other `Rewriter`s inherit from that class. To introduce SSA, the passes
+# insert the children, that should become operands, before the matched op, then
+# use the `added_operations_before` functionality of the rewriter to link the
+# operands to the new operation.
+
+
+@dataclass
+class RelAlgRewriter(RewritePattern):
+
+  def convert_datatype(self, type_: RelAlg.DataType) -> RelSSA.DataType:
+    if isinstance(type_, RelAlg.String):
+      return RelSSA.String.get(type_.nullable)
+    if isinstance(type_, RelAlg.Int32):
+      return RelSSA.Int32()
+    raise Exception(
+        f"datatype conversion not yet implemented for {type(type_)}")
+
+  def lookup_type_in_schema(self, name: str,
+                            bag: RelSSA.Bag) -> Optional[RelSSA.DataType]:
+    """
+    Looks up the type of name in the schema of bag.
+    """
+    for s in bag.schema.data:
+      if s.elt_name.data == name:
+        return s.elt_type
+    return None
+
+  def lookup_type_in_parent_op(self, name: str, parent_op: Operation):
+    """
+    Crawls through all parent_ops until reaching either a ModuleOp, in which
+    case the lookup failed or reaching an operation with an input bag, that the
+    type can be looked up in.
+    """
+    if isinstance(parent_op, ModuleOp):
+      raise Exception(f"element not found in parent schema: {name}")
+    if isinstance(parent_op, RelAlg.Select):
+      type_ = self.lookup_type_in_schema(name,
+                                         parent_op.input.op.results[0].typ)
+      if type_:
+        return type_
+      raise Exception(f"element not found in parent schema: {name}")
```

Comment:
Why does one of the two functions raise an exception and the other one return `None`? The names have the same structure -- that suggests the same behavior, no?

Reply:
I imagined that there might be reasons why we want to get a None from the first function later on, whereas I was sure what the exact behavior of the second function should be. I guess the return type hints at what will happen, but I agree this should be more expicit.

I renamed the second to `find_...` s.t. the naming does not suggest the same behavior as much as before.